### PR TITLE
MXNET-873 - Bring Clojure Package Inline with New DataDesc and Layout in Scala Package

### PR DESCRIPTION
**Pull request by gigasquid**
## Description ##
The Scala package has updated the DataDesc to include Layout. The Clojure package has been updated to move inline with it.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- A layout namespace has been created and used in DataDesc
- The Clojure package no longer infers the layout from the shape. It handles it the same way as the Scala package by having it be undefined
- The original `module.fit` interop code has been restored. The work arounds required by the old DataDesc has been resolved with this PR https://github.com/apache/incubator-mxnet/pull/11844
- Most of the examples have moved to use `provide-data-desc` instead of `provide-data`
- Formatting fixes
- Tweak to Char RNN example to run for less epochs before showing the pre-loaded result

### CODE REVIEW

**Comment by gigasquid** [comment]
@lanking520 please give a look when you get a chance

File: contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj [hunk 1]
```diff
@@ -92,7 +92,7 @@
 
 (comment
 
-  (predict "https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/python/predict_image/cat.jpg")
+  (predict "https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/python/predict_image/cat.jpg" true)
```
Comment (lanking520) [inline contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj]: What does this `true` mean in here?

File: contrib/clojure-package/src/org/apache/clojure_mxnet/symbol.clj [hunk 2]
```diff
@@ -144,7 +144,7 @@
    which must be known from the rest of the net."
   ([start {:keys [step repeat dtype]
            :or {step (float 1) repeat (int 1) dtype base/MX_REAL_TYPE}
-          :as opts}]
+           :as opts}]
```
Comment (lanking520) [inline contrib/clojure-package/src/org/apache/clojure_mxnet/symbol.clj]: space issue...

**Comment by gigasquid** [comment]
Thanks for the feedback @lanking520 😸

**Comment by lanking520** [comment]
@gigasquid Sorry about not being so helpful there... I will try to learn Clojure grammer and review it again... Generally, the Scala DataDesc want user to pass in as DataDesc type instead of ListMap. Then most of the iterators needs to change a little bit based on this change.

**Comment by gigasquid** [comment]
Thanks @lanking520 I did change the NDarrayIter to reflect this change. You can probably see it easiest in the test case https://github.com/apache/incubator-mxnet/pull/12387/files#diff-93817a3c2dc0abcdbb709a400bddc997R194

Please feel free to ping me with any questions.

File: contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj (hunk 1 above)
Comment (gigasquid) [reply contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj]: It's for the example - it controls whether the image url displays in a popup for you view 
https://github.com/apache/incubator-mxnet/blob/6fd245d1ce28a7bc09b83aeb2a1233f085210037/contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj#L42

File: contrib/clojure-package/src/org/apache/clojure_mxnet/io.clj [hunk 3]
```diff
@@ -213,14 +245,24 @@
...
                label-name "label"}}]
-   (new NDArrayIter
-        (util/vec->indexed-seq data)
-        (if label (util/vec->indexed-seq label) (util/empty-indexed-seq))
-        (int data-batch-size)
-        shuffle
-        last-batch-handle
-        data-name
-        label-name))
+   (let [specify-data-desc? (map? data)]
```
Comment (benkamphaus) [inline contrib/clojure-package/src/org/apache/clojure_mxnet/io.clj]: nit: is this just in the `let` for a self-documenting name/rationale?

File: contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj (hunk 1 above)
Comment (benkamphaus) [reply contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj]: nit: not a big deal since it's a helper fn in an example, but it'd make sense to use an options map or kw arg (`{:display true}` or `:display true`). [outside this PR though, also]

Comment (gigasquid) [reply contrib/clojure-package/examples/pre-trained-models/src/pre_trained_models/predict_image.clj]: yes agree

File: contrib/clojure-package/src/org/apache/clojure_mxnet/io.clj (hunk 3 above)
Comment (gigasquid) [reply contrib/clojure-package/src/org/apache/clojure_mxnet/io.clj]: I think it can be dropped - thanks

**Comment by lanking520** [comment]
@nswamy @yzhliu Please take a look since it's been for a while

File: contrib/clojure-package/src/org/apache/clojure_mxnet/io.clj [hunk 4]
```diff
@@ -213,14 +245,23 @@
...
                label-name "label"}}]
-   (new NDArrayIter
-        (util/vec->indexed-seq data)
-        (if label (util/vec->indexed-seq label) (util/empty-indexed-seq))
-        (int data-batch-size)
-        shuffle
-        last-batch-handle
-        data-name
-        label-name))
+   (if (map? data)
```
Comment (yzhliu) [inline contrib/clojure-package/src/org/apache/clojure_mxnet/io.clj]: does it mean if data is map? what if it is not? I don't know clojure well, just want to make sure it is intended.

Comment (gigasquid) [reply contrib/clojure-package/src/org/apache/clojure_mxnet/io.clj]: Yes - it checks to see if the data is a map, if so, it is in the form of having a DataDesc associated with it and will be dispatched to the correct Java function signature and with scala interop. If it is not a map, it will dispatch to the original Java function signature without DataDesc. 

The argument checking for the correct data structures can be improved by using core.spec in Clojure. It adds gradual type checking. It is in use in the module api, but it hasn't been added in yet in this namespace. I added a line item in the TODO page for the Clojure package to capture it for later improvement work.

**Comment by gigasquid** [comment]
Thanks @yzhliu for the feedback 😸 

Unless anyone has any objections, I will go ahead a merge shortly.
